projects/com.oracle.truffle.llvm.nodes.impl/src/com/oracle/truffle/llvm/nodes/impl/memory/load/LLVMDirectLoadNode.java

```diff
@@ -91,6 +93,22 @@ public LLVMFunctionDescriptor executeAddress(LLVMAddress addr) {
         public LLVMAddress executeAddress(LLVMAddress addr) {
             return LLVMMemory.getAddress(addr);
         }
+
+        // The following specializations are simpler than supporting arbitrary foreign objects here
+
+        @Specialization
+        public Object executeAddress(ManagedMallocObject addr) {
+            return addr.get(0);
+        }
+
+        @Specialization(guards = "objectIsManagedMalloc(addr)")
```

Comment:
Yes the first is the case of dereferencing a managed malloc with either `*value` or `value[0]` when the `0` is constant, and the second is the case of an arbitrary index or an inner pointer.